cli/auth.go

```diff
@@ -87,21 +88,113 @@ type Token struct {
 	TokenURL     string    `json:"token_url"`
 	ClientID     string    `json:"client_id"`
 
-	User UserData `json:"user_data"`
+	User userData `json:"user_data"`
 }
 
-// IsExpired returns true if the token is expired.
-func (t *Token) IsExpired() bool {
+// LoginAction is the corresponding Action for 'login'.
+func LoginAction(c *cli.Context) error {
+	client, err := newAppClient(c)
+	if err != nil {
+		return err
+	}
+
+	loggedInMessage := func(t *token, alreadyLoggedIn bool) {
+		already := "already "
+		if !alreadyLoggedIn {
+			already = ""
+			viamLogo(c.App.Writer)
+		}
+
+		fmt.Fprintf(c.App.Writer, "%slogged in as %q, expires %s\n", already, t.User.Email,
+			t.ExpiresAt.Format("Mon Jan 2 15:04:05 MST 2006"))
+	}
+
+	if client.conf.Auth != nil && !client.conf.Auth.isExpired() {
+		loggedInMessage(client.conf.Auth, true)
+		return nil
+	}
+
+	var t *token
+	if client.conf.Auth != nil && client.conf.Auth.canRefresh() {
+		t, err = client.authFlow.refreshToken(client.c.Context, client.conf.Auth)
+		if err != nil {
+			utils.UncheckedError(client.logout())
+			return err
+		}
+	} else {
+		t, err = client.authFlow.login(client.c.Context)
+		if err != nil {
+			return err
+		}
+	}
+
+	// write token to config.
+	client.conf.Auth = t
+	if err := storeConfigToCache(client.conf); err != nil {
+		return err
+	}
+
+	loggedInMessage(client.conf.Auth, false)
+	return nil
+}
+
+// PrintAccessTokenAction is the corresponding Action for 'print-access-token'.
+func PrintAccessTokenAction(c *cli.Context) error {
+	client, err := newAppClient(c)
+	if err != nil {
+		return err
+	}
+
+	if client.conf.Auth == nil || client.conf.Auth.isExpired() {
+		return errors.New("not logged in. run \"login\" command")
+	}
+
```

Comment:
[opt/suggestion] should we call `ensureLoggedIn` here? That would refresh the token if it is expired rather than telling the user to re log in.

Reply:
Great idea! Added. I left the current `WhoAmI` "not logged in" warning, as I feel like that's more of a warning and not an error.